qiskit/transpiler/target.py

```diff
@@ -436,48 +441,77 @@ def update_from_instruction_schedule_map(self, inst_map, inst_name_map=None, err
                     else:
-                        val.duration = None
-                    if error_dict is not None:
-                        error_inst = error_dict.get(inst)
-                        if error_inst:
-                            error = error_inst.get(qarg)
-                            val.error = error
-                        else:
-                            val.error = None
-                    else:
-                        val.error = None
-                out_props[qarg] = val
-            if inst not in self._gate_map:
-                if inst_name_map is not None:
-                    self.add_instruction(inst_name_map[inst], out_props, name=inst)
+                        duration = None
+                    props = InstructionProperties(
+                        duration=duration,
+                        calibration=entry,
+                    )
                 else:
-                    raise ValueError(
-                        "An inst_name_map kwarg must be specified to add new "
-                        "instructions from an InstructionScheduleMap"
+                    if props is None:
+                        # Edge case. Calibration is backend defined, but this is not
+                        # registered in the backend target. Ignore this entry.
+                        continue
+                try:
+                    # Update gate error if provided.
+                    props.error = error_dict[inst_name][qargs]
+                except (KeyError, TypeError):
+                    pass
+                out_props[qargs] = props
+            if not out_props:
+                continue
+            if inst_name not in self._gate_map:
+                if inst_name_map is None:
+                    inst_name_map = get_standard_gate_name_mapping()
+                if inst_name in inst_name_map:
+                    # Remove qargs with length that doesn't match with instruction qubit number
+                    inst_obj = inst_name_map[inst_name]
+                    normalized_props = {}
+                    for qargs, prop in out_props.items():
+                        if len(qargs) != inst_obj.num_qubits:
+                            continue
+                        normalized_props[qargs] = prop
+                    self.add_instruction(inst_obj, normalized_props, name=inst_name)
+                else:
+                    # Assumes qubit number and parameter names are consistent
+                    qargs0 = inst_map.qubits_with_instruction(inst_name)[0]
+                    try:
+                        qargs0 = tuple(qargs0)
+                    except TypeError:
+                        qargs0 = (qargs0,)
+                    cal = getattr(out_props[qargs0], "_calibration")
+                    inst_obj = Gate(
+                        name=inst_name,
+                        num_qubits=len(qargs0),
+                        params=list(map(Parameter, cal.get_signature().parameters.keys())),
```

Comment:
We don't want to fall back to use the standard gates in this case too? I feel like the user expectation is if you did something like `cx` and also specified a custom name mapping that it would use `CXGate`.

Reply:
If I understand correctly you pointed out the case that user provides `inst_name_map` only for the custom gate name? (then `cx` doesn't appear in the mapping and context-less `Gate` object is assigned). I updated the logic so that it always takes the standard gate defs in [bdb790a](https://github.com/Qiskit/qiskit-terra/pull/9587/commits/bdb790af888455c2f31d5fa7d111a9d697798796). Please let me know if I misunderstand your point.

Reply:
Yeah, that was my concern [bdb790a](https://github.com/Qiskit/qiskit-terra/pull/9587/commits/bdb790af888455c2f31d5fa7d111a9d697798796) should fix it, thanks.